### DISCUSSION TITLE
How does this 3D canvas mockup script by GeeMack works?

### DISCUSSION THREAD

**Discussion by playgamemy** [ImageMagick/ImageMagick#4231] [Help] [answered]
I am new to IM I found a post in old discussion forum and found a script answered by GeeMack to turn image into 3D canvas mockup.
[](https://legacy.imagemagick.org/discourse-server~/viewtopic59e8.html?p=172882)
I ran the script on my PHP server with exec, and it works the way I need. Now I want to modify the script so it fits my needs, but I cannot make head or tail of this script, mostly after the first few lines after images were deleted from the sequence.

`convert logo: -bordercolor none -border 1 -crop 64x0 \( -clone 0 \
 -distort perspective "0,0 0,32 64,0 64,0 0,480 0,448 64,480 64,480" \
 -modulate 90 -write mpr:edge +delete \) -delete 0 +append \
 -distort perspective "0,0 0,0 0,480 0,480 576,0 576,32 576,480 576,448" \
 mpr:edge +swap +append -border 32 \( +clone -colorize 100 \
 -distort affine "10,0 0,10" -blur 0x10 \) +insert -composite result.png`

My understanding is it loaded the IM logo image as its first image in the sequence, added a transparent border of 1 pixel to all side, cropped its left edge out, then clone the cropped edge, distort perspective, lowered brightness, and write it into memory with the name 'edge', then deleted the clone.

What happen afterwards puzzle me, in my understanding there is only one image in the sequence at this point (the cropped edge without perspective distorted), ... and it deleted the only image? then append what horizontally when there are no images in the sequence?

If I delete the rest of the script from the first '`+append`', then the script won't run, so I really don't know how to troubleshoot to answer my own question.

If I delete the `-bordercolor none -border 1` then everthing fall apart. Maybe the -border 1 operator actually create a new image in the sequence? That does not make sense to me though.... In fact, I do not understand why a transprarent border of 1 pixel is needed in the first place? What does it achieve?

Can anyone help me decipher this?

**Comment by snibgo** [comment, marked as answer]
You misunderstand "-crop 64x0". See https://imagemagick.org/script/command-line-options.php#crop

There are no offsets, so this creates multiple images. All (possibly except for the last) are 64 pixels wide. From the documentation:

> If the x and y offsets are omitted, a set of tiles of the specified geometry, covering the entire input image, is generated. The rightmost tiles and the bottom tiles are smaller if the specified geometry extends beyond the dimensions of the input image.

So "-delete 0" removes the left-most tile, and "+append" glues the others back together again.

Incidentally, when quoting "a post in old discussion forum" please link to it, so we can see the context.

**Reply by playgamemy** [reply]
@snibgo Thank you! That make so much more sense now! How about the very first 1pixel transparent border? What does it do? Why the script go haywire if I remove that part then?
I understand the second parenthesis clone is to generate the shadow, but what is +insert? the documnet does not make much sense "inserting the last image at the end of the current sequence", if it is the last image... is it not already the last of the current sequence?
(wait a minute, I think I understand +insert put the shadow underneaththe image before composite the 2 images together... so every sequent image loaded is placed on top... so technically after reading 'first' image, and reading 'second' image next actually put them in a sequence of 'second, first' but not 'first, second'?)

**Comment by snibgo** [comment]
When I need to understand a script, I write it out like this:
```
magick \
  logo: \
  -bordercolor "#88f" -border 1 -crop 64x0 \
  \( -clone 0 \
     -distort perspective "0,0 0,32 64,0 64,0 0,480 0,448 64,480 64,480" \
     -modulate 90 -write mpr:edge \
+write x0.png \
     +delete \
  \) \
  -delete 0 +append \
+write x1.png \
  -distort perspective "0,0 0,0 0,480 0,480 576,0 576,32 576,480 576,448" \
+write x2.png \
  mpr:edge +swap +append -bordercolor "#22f" -border 32 \
+write x3.png \
  \( +clone -colorize 100 \
     -distort affine "10,0 0,10" -blur 0x10 \
+write x4.png \
  \) \
  +insert \
  -composite \
  result.png
```
Note that I have inserted operations like "+write x0.png". This writes intermediate results to files that I can then view. For example, I can clarly see that x4 is a black blurred copy that will be used as a shadow. And this needs to be behind, not in front of, whatever it is a shadow of. "+insert" changes the order of images so whatever was on top goes behind all the others.

I have also changed colours I can't see ("none") into colours I can see. The light blue is the first border. The dark blue is the second.
![result](https://user-images.githubusercontent.com/33812368/133539755-dda19baa-2265-401f-94ac-1cd034e005a8.png)

We can see the first border, which started as one pixel wide, has been stretched by both perspective distortions. The borders are solid colours, so they conceal the shadow. If we had no border, then the edge pixels (which are white) would be stretched instead, so they would also conceal the shadow, and the sloping edges would not be visible.

**Reply by fmw42** [reply]
I do not see any difference between x3.png and result.png. What is the intent of the -distort affine and is there some mistake here that causes x3.png and result.png to be the same. x4.png is a black image and is apparently covered over. I assume it is intended for a background shadow. But it does not show. Is that because the background was set to blue?

**Reply by fmw42** [reply]
OK. If I change the blue background to "none" a shadow appears. So all is fine.

**Reply by snibgo** [reply]
>  I assume it is intended for a background shadow. But it does not show. Is that because the background was set to blue?

Yes. I used (opaque) blue to explain what the first (transparent) border is used for.
